### PR TITLE
feat!: use arrays to represent commands, new `resolveCommand` function

### PR DESCRIPTION
**Pull request by userquin**
### Description

This PR converts every command in `COMMANDS` to command and argument pairs.

And provide a new utility, `constructCommand`, to make the resolution easier.

### Linked Issues
https://github.com/antfu-collective/ni/pull/231#discussion_r1718864647

### Additional context

I do my best to keep types similar to `COMMANDS` in `agents.ts`, I guess it is fine using `satisfies` in the `buildCommands` return type.

### CODE REVIEW

File: src/commands.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import type { Agent } from './agents'
+import { COMMANDS } from './agents'
```
Comment (antfu) [inline src/commands.ts:2]: Should we do that on the definition side?

**Comment by userquin** [comment]
don't merge yet, I'm going to add an utility helper to allow return the CommandType in all commands, that's, all new agents commands actions will be functions like `npm.run`: will also handle `{0}`

**Comment by userquin** [comment]
This is the current  `ni` behavior, generated commands here with same behavior:

imagen

**Comment by benmccann** [comment]
it would be helpful to update the readme showing how to use the new functionality

File: README.md [hunk 2]
```diff
@@ -58,10 +58,23 @@ const pm = await detect()
...
+Almost every command from `COMMANDS` will accept arguments, if you want to get the command with custom arguments, you can use `AGENT_COMMANDS` instead. For example, if you want to show how to install `my-awesome-lib` dependency:
+```js
+import { detect } from 'package-manager-detector'
+import { AGENT_COMMANDS } from 'package-manager-detector/agents'
+
+const pm = await detect()
+if (!pm)
+  throw new Error('Could not detect package manager')
+
+const command = AGENT_COMMANDS[pm]['add']
```
Comment (benmccann) [inline README.md:74]: I wonder if we should make this `.add` instead and remove the `lint` config allowing this? I'm happy either way

Comment (userquin) [reply README.md]: done

File: src/agents.ts [hunk 3]
```diff
@@ -94,6 +97,92 @@ export const COMMANDS = {
...
+
+function buildCommands(agent: Agent, commands: typeof COMMANDS.npm) {
+  const interactive = commands['upgrade-interactive'] as string | null
+  return {
+    'agent': commands.agent,
+    'run': typeof commands.run === 'function'
+      ? npmRunCommand(agent.includes('@') ? agent.split('@')[0] : agent)
+      : buildCommand(commands.run as string),
+    'install': buildCommand(commands.install),
+    'frozen': buildCommand(commands.frozen),
```
Comment (antfu) [inline src/agents.ts:168]: Build command feels over complicated, can't we update the static map above to have arrays directly? Like

```ts
const yarn = {
  'agent': ['yarn', 0],
  'run': ['yarn', 'run', 0],
   // ...
}
```

**Comment by antfu** [comment]
I refactored it a bit. So the `COMMANDS` is now an object of arrays. Which means it would be a breaking change. I also provided a utility `constructCommand` to make it easier to construct.

@benmccann could you help to confirm how much it would affect your usage? If not much I am thinking to have it as 0.2 and then update the usage side on `ni`

**Review by antfu — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -52,14 +52,14 @@ This package includes package manager agents and their corresponding commands fo
...
+import { COMMANDS, constructCommand } from 'package-manager-detector/agents'
 
 const pm = await detect()
-if (!pm)
+if (!pm.agent)
   throw new Error('Could not detect package manager')
 
-const command = COMMANDS[pm.agent]['frozen']
-console.log(`Detected the ${pm.agent} package manager. You can run a frozen install with ${command}`)
+const commandArgs = constructCommand(COMMANDS[pm.agent].add, ['@antfu/ni']) // ['pnpm', 'add', '@antfu/ni']
```
Comment (benmccann) [inline README.md:61]: I wonder if this should return something like:
```
{
  cmd: 'pnpm',
  args: ['add', '@antfu/ni']
}
```

Most of the time when you're executing the command and args need to be handled separately. E.g. https://www.npmjs.com/package/tinyexec shows `x('ls', ['-l'])`

Comment (antfu) [reply README.md]: Good idea, updated

**Pull request merged**